### DISCUSSION TITLE
Drawing 3D scene to a canvas with WGPU

### DISCUSSION THREAD

**Discussion by junglie85** [emilk/egui#1661] [Q&A]
I've been playing around trying to render a scene in my toy game engine to an image with some success using WGPU as the render backend. However, having been kindly pointed towards the custom 3D painter example, I see that a canvas is probably a better option. Following the example, I add a canvas and set a callback, but the callback never gets called:

```
egui::CentralPanel::default().show(&egui_ctx, |ui| {
        egui::Frame::canvas(ui.style()).show(ui, |ui| {
            let (rect, response) =
                ui.allocate_exact_size(egui::Vec2::splat(300.0), egui::Sense::drag());
            dbg!(response);

            let callback = egui::PaintCallback {
                rect,
                callback: std::sync::Arc::new(move |_info, _render_ctx| {
                    panic!("Callback got called");
                }),
            };
            ui.painter().add(callback);
        });
    });
```

I'm not familiar enough with egui to know where to start looking to diagnose why the callback isn't being called. If someone would be willing to point me in the right direction, I'd be very grateful.

**Comment by KentaTheBugMaker** [comment]
now paint callback on wgpu backend not supported.

**Reply by junglie85** [reply]
What’s required to get the paint callback supported?

**Reply by emilk** [reply]
Someone willing to make a PR :)

https://github.com/emilk/egui/blob/master/egui-wgpu/src/renderer.rs#L499

I'm not sure how to best structure it (because I haven't thought about it).

If you'd like to work on it, it would be greatly appreciated!

**Reply by zicklag** [reply]
No promises, but I think I'm going to need this, too, so I'm going to take a look at it.

**Comment by zicklag** [comment]
I'm working on a pull request to support this and I got a WGPU triangle working in my app!

![image](https://user-images.githubusercontent.com/25393315/170522782-446a94bf-b13b-4521-99e9-c6dd90a902a4.png)

I should probably try porting the WGPU cube example to make sure it can take handle more than just a triangle, but the proof-of-concept is going well and I don't see it taking a ton of effort to get mergable.

---

Also, **quick question for anybody who has more experience with WGPU than I do**: is it a big limitation to require custom render callbacks to write to the same `RenderPass` used by the rest of the GUI?

I figure if you need to do your own multi-pass rendering or something like that, you can just render to a texture and then render that texture into the main egui `RenderPass`.

This allows us to keep the viewport and clipping rects set automatically by the backend. It also requires less work on the backend to make this work.

If anybody thinks this might be a problem, let me know and I can look into alternatives.

I'll be pushing a draft pull request soon!

**Reply by zicklag** [reply]
Produced a WGPU equivalent to the glow demo:
![Peek 2022-05-26 13-48](https://user-images.githubusercontent.com/25393315/170557017-49ca15f6-5d95-49fb-a8cf-c1a0ee5e07f1.gif)

Also refined the API a little bit and I'm just about happy with it.

**Reply by zicklag** [reply]
Got a pull request up! #1684 

And here's the usage example ( Linking to a Gist so the link doesn't break if I have to push more commits to the PR ):

https://gist.github.com/zicklag/b9c1be31ec599fd940379cecafa1751b

**Reply by emilk** [reply]
And it's merged 🥳

Thanks!

**Reply by zicklag** [reply]
Glad I could help, and thank you! I've been enjoying Egui. :)
